sql/core/src/main/scala/org/apache/spark/sql/DataFrameStatFunctions.scala

```diff
@@ -375,8 +375,11 @@ final class DataFrameStatFunctions private[sql](df: DataFrame) {
     import org.apache.spark.sql.functions.{rand, udf}
     val c = Column(col)
     val r = rand(seed)
-    val f = udf { (stratum: Any, x: Double) =>
-      x < fractions.getOrElse(stratum.asInstanceOf[T], 0.0)
+    // Hack to get around the fact that type T is Any and we can't use a UDF whose arg
+    // is Any. Convert everything to a string rep.
```

Comment:
I'm not sure this is safe to do. Maybe `a == b` is different from `a.toString == b.toString`.

Reply:
I feel udf using `Any` as input type is rare but a valid use case. Sometimes they just want to accept any input type.

How about we create a few `udfInternal` methods that takes `Any` as inputs? e.g.
```
def udfInternal[R: TypeTag](f: Function1[Any, R]): UserDefinedFunction = {
  val ScalaReflection.Schema(dataType, nullable) = ScalaReflection.schemaFor[R]
  val udf = UserDefinedFunction(f, dataType, Nil)
  if (nullable) udf else udf.asNonNullable()
}
```

Reply:
You are right that the change I made here is not bulletproof. Unfortunately there are several more problems like this. Anywhere there's a UDF on `Row` it now fails, and workarounds are ugly.

I like your idea, let me work on that. Because the alternative I've been working on is driving me nuts.